File: 368-largest-divisible-subset/largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int> &nums) {
        nums.push_back(1);
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, -1));
        solve(1, 0, nums, dp, n);
        vector<int> ans;
        build(1, 0, nums, dp, n, ans);
        return ans;
    }

    int solve(int idx, int prv, vector<int> &nums, vector<vector<int>> &dp, int &n) {
        if (idx == n)
            return 0;
        int &ret = dp[idx][prv];
        if (~ret)
            return ret;
        ret = 0;
        if (nums[idx] % nums[prv] == 0)
            ret = solve(idx + 1, idx, nums, dp, n) + 1;
        ret = max(ret, solve(idx + 1, prv, nums, dp, n));
        return ret;
    }

    void build(int idx, int prv, vector<int> &nums, vector<vector<int>> &dp, int &n, vector<int> &ans) {
        if (idx == n)
            return;
        int &ret = dp[idx][prv];
        if (nums[idx] % nums[prv] == 0 && ret == solve(idx + 1, idx, nums, dp, n) + 1) {
            ans.push_back(nums[idx]);
            build(idx + 1, idx, nums, dp, n, ans);
            return;
        }
        build(idx + 1, prv, nums, dp, n, ans);
    }
};

File: 368-largest-divisible-subset/largest-divisible-subset.py
from math import sqrt

class Solution(object):
    def largestDivisibleSubset(self, nums):
        nums = sorted(nums)
        prev_divisor = {}
        for i in range (0, len(nums)):
            prev_divisor[i] = [i, 1]
            for j in range(0, i):
                if nums[i] % nums[j] == 0 and prev_divisor[i][1] < prev_divisor[j][1] + 1:
                    prev_divisor[i][0], prev_divisor[i][1] = j, prev_divisor[j][1] + 1
        mx_sz, idx = 0, 0
        for i in range (0, len(nums)):
            if mx_sz < prev_divisor[i][1]:
                mx_sz, idx = prev_divisor[i][1], i
        ans = [nums[idx]]
        while prev_divisor[idx][0] != idx:
            ans.append(nums[prev_divisor[idx][0]])
            idx = prev_divisor[idx][0]
        return ans
        